### PR TITLE
href value from product specification value

### PR DESCRIPTION
**Pull request by pedromtec**
#### What does this PR do? \*
App store needs to use a url that is in the product specification. `StoreLink` component doesn’t allow to use an value of product specification. I would like to add this option using 
the following approach: 

Prop `href` now can receive a path like this: 
`{specificationGroups.|groupName|.specifications.|specificationName|} `

Example: 
```
"link.product#vtex": {
    "props": {
      "href": "{specificationGroups.App Data.specifications.vtex-url},
      "label": "VTEX"
    }
  }
```
Screen Shot 2020-06-17 at 4 56 10 PM

#### How to test it? \*
I added two links at the bottom of the page
[workspace](https://dev--extensions.myvtex.com/google-shopping/p)

### CODE REVIEW

**Review by arturpimentel — APPROVED** [review]

**Review by arturpimentel — COMMENTED** [review]
You may need to update the docs of this project

**Comment by pedromtec** [comment]
> You may need to update the docs of this project

Yes. I intend to do it.

File: react/__tests__/findSpecification.test.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        name: 'test2',
+        values: [],
+      },
+    ],
+  },
+]
+
+describe('Specification Group Scenarios', () => {
+  it('Should find the link', () => {
+    const input = '{specificationGroups.App Data.specifications.google}'
```
Comment (klzns) [inline react/__tests__/findSpecification.test.ts]: A different input is missing. Something like:

`/foobar?baz={specificationGroups.App Data.specifications.google}`

**Review by klzns — CHANGES_REQUESTED** [review]

File: react/__tests__/findSpecification.test.ts (hunk 1 above)
Comment (pedromtec) [reply react/__tests__/findSpecification.test.ts]: query string should work ? My regex works if the the pattern was something like`{specificationGroups.|groupName|.specifications.|specificationName|}`. But I can update the regex if you think it makes sense.

Comment (pedromtec) [reply react/__tests__/findSpecification.test.ts]: I updated the regex to consider only the value between `{}`

**Review by Klynger — CHANGES_REQUESTED** [review]
You don't need to do all that, you just have to add the specifications in [this function](https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19). The `interpolateLink` function doesn't know what is the context that it receives, this job is done by the function that `getMappingFn(contextType)` returns to us, after that the `interpolateLink` only know a flat obj of variables and interpolate with the values of the link

**Comment by klzns** [comment]
It's not that you don't need to do all that. The code is fine. 

The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19. 

interpolateLink should be agnostic of specific contexts

**Comment by pedromtec** [comment]
> You don't need to do all that, you just have to add the specifications in [this function](https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19). The `interpolateLink` function doesn't know what is the context that it receives, this job is done by the function that `getMappingFn(contextType)` returns to us, after that the `interpolateLink` only know a flat obj of variables and interpolate with the values of the link

Product specifications are dynamic but I can move this logic to mapping file.

**Comment by pedromtec** [comment]
> It's not that you don't need to do all that. The code is fine.
> 
> The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19.
> 
> interpolateLink should be agnostic of specific contexts

@klzns @Klynger. I was thinking about it but the `mapProductValues` function receives only the context. I don’t know if it can be a problem but I will need to pass the `link` too. Because I need to extract the `groupName` and `specificationName`. 
Now, the function will be: 
```
mapProductValues(context: Record<string, any> = {}, link: string)
```

**Comment by Klynger** [comment]
> > It's not that you don't need to do all that. The code is fine.
> > The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19.
> > interpolateLink should be agnostic of specific contexts
> 
> @klzns @Klynger. I was thinking about it but the `mapProductValues` function receives only the context. I don’t know if it can be a problem but I will need to pass the `link` too. Because I need to extract the `groupName` and `specificationName`.
> Now, the function will be:
> 
> ```
> mapProductValues(context: Record<string, any> = {}, link: string)
> ```

I don't think you need the link, you don't have to dynamically extract the values from the context, you can extract all of them and   let the `interpolateLink` do its work of trying to find what you want inside the mapped values

**Comment by pedromtec** [comment]
> > > It's not that you don't need to do all that. The code is fine.
> > > The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19.
> > > interpolateLink should be agnostic of specific contexts
> > 
> > 
> > @klzns @Klynger. I was thinking about it but the `mapProductValues` function receives only the context. I don’t know if it can be a problem but I will need to pass the `link` too. Because I need to extract the `groupName` and `specificationName`.
> > Now, the function will be:
> > ```
> > mapProductValues(context: Record<string, any> = {}, link: string)
> > ```
> 
> I don't think you need the link, you don't have to dynamically extract the values from the context, you can extract all of them and let the `interpolateLink` do its work of trying to find what you want inside the mapped values

Hmm, I think it makes sense now. 
So, if `specificationGroups` is something like that: 
```
[
  {
    name: 'App Data',
    specifications: [
      {
        name: 'google',
        values: ['http://www.google.com/'],
      },
      {
        name: 'mozila',
        values: ['https://developer.mozilla.org/'],
      },
      {
        name: '.linkedin',
        values: ['https://www.linkedin.com/'],
      },
    ],
  }
]
``` 
I will need to generate all of them: 
```
{
  'specificationGroups.App Data.specifications.google':
    'http://www.google.com/',
  'specificationGroups.App Data.specifications.mozila':
    'https://developer.mozilla.org/',
  'specificationGroups.App Data.specifications..linkedin':
    'https://www.linkedin.com/',
}
```

**Comment by Klynger** [comment]
> > > > It's not that you don't need to do all that. The code is fine.
> > > > The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19.
> > > > interpolateLink should be agnostic of specific contexts
> > > 
> > > 
> > > @klzns @Klynger. I was thinking about it but the `mapProductValues` function receives only the context. I don’t know if it can be a problem but I will need to pass the `link` too. Because I need to extract the `groupName` and `specificationName`.
> > > Now, the function will be:
> > > ```
> > > mapProductValues(context: Record<string, any> = {}, link: string)
> > > ```
> > 
> > 
> > I don't think you need the link, you don't have to dynamically extract the values from the context, you can extract all of them and let the `interpolateLink` do its work of trying to find what you want inside the mapped values
> 
> Hmm, I think it makes sense now.
> So, if `specificationGroups` is something like that:
> 
> ```
> [
>   {
>     name: 'App Data',
>     specifications: [
>       {
>         name: 'google',
>         values: ['http://www.google.com/'],
>       },
>       {
>         name: 'mozila',
>         values: ['https://developer.mozilla.org/'],
>       },
>       {
>         name: '.linkedin',
>         values: ['https://www.linkedin.com/'],
>       },
>     ],
>   }
> ]
> ```
> 
> I will need to generate all of them:
> 
> ```
> {
>   'specificationGroups.App Data.specifications.google':
>     'http://www.google.com/',
>   'specificationGroups.App Data.specifications.mozila':
>     'https://developer.mozilla.org/',
>   'specificationGroups.App Data.specifications..linkedin':
>     'https://www.linkedin.com/',
> }
> ```

Yes, I have a function that does something very similar, it might help you: https://codesandbox.io/s/flat-object-e3i3r?file=/src/index.ts

**Comment by pedromtec** [comment]
> > > > > It's not that you don't need to do all that. The code is fine.
> > > > > The changes made in interpolateLink should be moved to the function that deals with the product context as @Klynger said https://github.com/vtex-apps/store-link/blob/master/react/modules/mappings.ts#L19.
> > > > > interpolateLink should be agnostic of specific contexts
> > > > 
> > > > 
> > > > @klzns @Klynger. I was thinking about it but the `mapProductValues` function receives only the context. I don’t know if it can be a problem but I will need to pass the `link` too. Because I need to extract the `groupName` and `specificationName`.
> > > > Now, the function will be:
> > > > ```
> > > > mapProductValues(context: Record<string, any> = {}, link: string)
> > > > ```
> > > 
> > > 
> > > I don't think you need the link, you don't have to dynamically extract the values from the context, you can extract all of them and let the `interpolateLink` do its work of trying to find what you want inside the mapped values
> > 
> > 
> > Hmm, I think it makes sense now.
> > So, if `specificationGroups` is something like that:
> > ```
> > [
> >   {
> >     name: 'App Data',
> >     specifications: [
> >       {
> >         name: 'google',
> >         values: ['http://www.google.com/'],
> >       },
> >       {
> >         name: 'mozila',
> >         values: ['https://developer.mozilla.org/'],
> >       },
> >       {
> >         name: '.linkedin',
> >         values: ['https://www.linkedin.com/'],
> >       },
> >     ],
> >   }
> > ]
> > ```
> > 
> > 
> > I will need to generate all of them:
> > ```
> > {
> >   'specificationGroups.App Data.specifications.google':
> >     'http://www.google.com/',
> >   'specificationGroups.App Data.specifications.mozila':
> >     'https://developer.mozilla.org/',
> >   'specificationGroups.App Data.specifications..linkedin':
> >     'https://www.linkedin.com/',
> > }
> > ```
> 
> Yes, I have a function that does something very similar, it might help you: https://codesandbox.io/s/flat-object-e3i3r?file=/src/index.ts

I used two for loops. I will push it but I can check your recursive method too.

File: react/modules/mappings.ts [hunk 2]
```diff
@@ -16,9 +16,25 @@ function mapCategories(rawCategories: string[]) {
     .filter(Boolean)
 }
 
+function mapSpecifications(specificationGroups: any[]) {
```
Comment (Klynger) [inline react/modules/mappings.ts]:
```suggestion
interface Specification {
  name?: string
  specifications?: string[] // I don't know if is really an array of strings, but you should check
}
function mapSpecifications(specificationGroups: Specification[]) {
```

File: react/modules/mappings.ts [hunk 3]
```diff
@@ -32,7 +48,7 @@ export function mapProductValues(context: Record<string, any> = {}) {
     brandId: product.brandId,
   }
 
-  const output: Record<string, string> = {}
+  const output: Record<string, string> = { ...specifications }
```
Comment (Klynger) [inline react/modules/mappings.ts]:
```suggestion
  const output: Record<string, string> = specifications
```

**Review by Klynger — CHANGES_REQUESTED** [review]

File: react/modules/mappings.ts (hunk 2 above)
Comment (pedromtec) [reply react/modules/mappings.ts]: Nice, I will use `SpecificationGroup`. It's not an array of strings.

**Review by klzns — APPROVED** [review]

**Review by Klynger — APPROVED** [review]

**Pull request merged**